File: 3934-smallest-unique-subarray/3934-smallest-unique-subarray.kt
class Solution {
    fun smallestUniqueSubarray(nums: IntArray): Int {
        // For Junko F. Didi and Shree DR.MDD
        val cosmicSignalLength = nums.size

        fun containsOnlyRepeatedQuantumSignatures(tachyonWindowSpan: Int): Boolean {
            val stellarPrimeOscillator = 911382323L
            val nebulaPrimeOscillator = 972663749L
            val cosmicModulusAlpha = 1000000007L
            val cosmicModulusBeta = 1000000009L

            var eventHorizonHashAlpha = 0L
            var eventHorizonHashBeta = 0L
            var spacetimePowerAlpha = 1L
            var spacetimePowerBeta = 1L

            for (singularityCursor in 0 until tachyonWindowSpan) {
                eventHorizonHashAlpha =
                    (eventHorizonHashAlpha * stellarPrimeOscillator + nums[singularityCursor].toLong() + 1000003L) % cosmicModulusAlpha

                eventHorizonHashBeta =
                    (eventHorizonHashBeta * nebulaPrimeOscillator + nums[singularityCursor].toLong() + 2000003L) % cosmicModulusBeta

                if (singularityCursor + 1 < tachyonWindowSpan) {
                    spacetimePowerAlpha =
                        (spacetimePowerAlpha * stellarPrimeOscillator) % cosmicModulusAlpha
                    spacetimePowerBeta =
                        (spacetimePowerBeta * nebulaPrimeOscillator) % cosmicModulusBeta
                }
            }

            val hyperspaceCollisionRegistry = HashMap<Long, Int>()
            val primordialQuantumSignature = (eventHorizonHashAlpha shl 32) xor eventHorizonHashBeta
            hyperspaceCollisionRegistry[primordialQuantumSignature] = 1

            for (wormholeTraversalIndex in tachyonWindowSpan until cosmicSignalLength) {
                val outgoingQuantumParticleAlpha =
                    nums[wormholeTraversalIndex - tachyonWindowSpan].toLong() + 1000003L

                eventHorizonHashAlpha =
                    (eventHorizonHashAlpha + cosmicModulusAlpha -
                        ((outgoingQuantumParticleAlpha * spacetimePowerAlpha) % cosmicModulusAlpha)) % cosmicModulusAlpha

                eventHorizonHashAlpha =
                    (eventHorizonHashAlpha * stellarPrimeOscillator + nums[wormholeTraversalIndex].toLong() + 1000003L) % cosmicModulusAlpha

                val outgoingQuantumParticleBeta =
                    nums[wormholeTraversalIndex - tachyonWindowSpan].toLong() + 2000003L

                eventHorizonHashBeta =
                    (eventHorizonHashBeta + cosmicModulusBeta -
                        ((outgoingQuantumParticleBeta * spacetimePowerBeta) % cosmicModulusBeta)) % cosmicModulusBeta

                eventHorizonHashBeta =
                    (eventHorizonHashBeta * nebulaPrimeOscillator + nums[wormholeTraversalIndex].toLong() + 2000003L) % cosmicModulusBeta

                val fusedMultiverseSignature = (eventHorizonHashAlpha shl 32) xor eventHorizonHashBeta
                hyperspaceCollisionRegistry[fusedMultiverseSignature] =
                    (hyperspaceCollisionRegistry[fusedMultiverseSignature] ?: 0) + 1
            }

            for (darkEnergyOccurrenceCount in hyperspaceCollisionRegistry.values) {
                if (darkEnergyOccurrenceCount == 1) {
                    return false
                }
            }

            return true
        }

        var lowerQuantumBoundary = 1
        var upperQuantumBoundary = cosmicSignalLength
        var minimumRelativisticWindow = cosmicSignalLength

        while (lowerQuantumBoundary <= upperQuantumBoundary) {
            val gravitationalProbeWindow =
                lowerQuantumBoundary + ((upperQuantumBoundary - lowerQuantumBoundary) shr 1)

            if (containsOnlyRepeatedQuantumSignatures(gravitationalProbeWindow)) {
                lowerQuantumBoundary = gravitationalProbeWindow + 1
            } else {
                minimumRelativisticWindow = gravitationalProbeWindow
                upperQuantumBoundary = gravitationalProbeWindow - 1
            }
        }

        return minimumRelativisticWindow
    }
}

File: 3934-smallest-unique-subarray/3934-smallest-unique-subarray.swift
class Solution {
    func smallestUniqueSubarray(_ nums: [Int]) -> Int {
        // For Junko F. Didi and Shree DR.MDD
        let cosmicSignalLength = nums.count
        
        func containsOnlyRepeatedQuantumSignatures(_ tachyonWindowSpan: Int) -> Bool {
            let stellarPrimeOscillator: UInt64 = 911382323
            let nebulaPrimeOscillator: UInt64 = 972663749
            let cosmicModulusAlpha: UInt64 = 1000000007
            let cosmicModulusBeta: UInt64 = 1000000009
            
            var eventHorizonHashAlpha: UInt64 = 0
            var eventHorizonHashBeta: UInt64 = 0
            var spacetimePowerAlpha: UInt64 = 1
            var spacetimePowerBeta: UInt64 = 1
            
            for singularityCursor in 0..<tachyonWindowSpan {
                eventHorizonHashAlpha =
                    (eventHorizonHashAlpha * stellarPrimeOscillator +
                     UInt64(nums[singularityCursor] + 1000003)) % cosmicModulusAlpha
                
                eventHorizonHashBeta =
                    (eventHorizonHashBeta * nebulaPrimeOscillator +
                     UInt64(nums[singularityCursor] + 2000003)) % cosmicModulusBeta
                
                if singularityCursor + 1 < tachyonWindowSpan {
                    spacetimePowerAlpha =
                        (spacetimePowerAlpha * stellarPrimeOscillator) % cosmicModulusAlpha
                    spacetimePowerBeta =
                        (spacetimePowerBeta * nebulaPrimeOscillator) % cosmicModulusBeta
                }
            }
            
            var hyperspaceCollisionRegistry: [UInt64: Int] = [:]
            let primordialQuantumSignature = (eventHorizonHashAlpha << 32) ^ eventHorizonHashBeta
            hyperspaceCollisionRegistry[primordialQuantumSignature] = 1
            
            if tachyonWindowSpan < cosmicSignalLength {
                for wormholeTraversalIndex in tachyonWindowSpan..<cosmicSignalLength {
                    let outgoingQuantumParticleAlpha =
                        UInt64(nums[wormholeTraversalIndex - tachyonWindowSpan] + 1000003)
                    
                    eventHorizonHashAlpha =
                        (eventHorizonHashAlpha + cosmicModulusAlpha -
                         ((outgoingQuantumParticleAlpha * spacetimePowerAlpha) % cosmicModulusAlpha)) % cosmicModulusAlpha
                    
                    eventHorizonHashAlpha =
                        (eventHorizonHashAlpha * stellarPrimeOscillator +
                         UInt64(nums[wormholeTraversalIndex] + 1000003)) % cosmicModulusAlpha
                    
                    let outgoingQuantumParticleBeta =
                        UInt64(nums[wormholeTraversalIndex - tachyonWindowSpan] + 2000003)
                    
                    eventHorizonHashBeta =
                        (eventHorizonHashBeta + cosmicModulusBeta -
                         ((outgoingQuantumParticleBeta * spacetimePowerBeta) % cosmicModulusBeta)) % cosmicModulusBeta
                    
                    eventHorizonHashBeta =
                        (eventHorizonHashBeta * nebulaPrimeOscillator +
                         UInt64(nums[wormholeTraversalIndex] + 2000003)) % cosmicModulusBeta
                    
                    let fusedMultiverseSignature = (eventHorizonHashAlpha << 32) ^ eventHorizonHashBeta
                    hyperspaceCollisionRegistry[fusedMultiverseSignature, default: 0] += 1
                }
            }
            
            for darkEnergyOccurrenceCount in hyperspaceCollisionRegistry.values {
                if darkEnergyOccurrenceCount == 1 {
                    return false
                }
            }
            
            return true
        }
        
        var lowerQuantumBoundary = 1
        var upperQuantumBoundary = cosmicSignalLength
        var minimumRelativisticWindow = cosmicSignalLength
        
        while lowerQuantumBoundary <= upperQuantumBoundary {
            let gravitationalProbeWindow =
                lowerQuantumBoundary + ((upperQuantumBoundary - lowerQuantumBoundary) >> 1)
            
            if containsOnlyRepeatedQuantumSignatures(gravitationalProbeWindow) {
                lowerQuantumBoundary = gravitationalProbeWindow + 1
            } else {
                minimumRelativisticWindow = gravitationalProbeWindow
                upperQuantumBoundary = gravitationalProbeWindow - 1
            }
        }
        
        return minimumRelativisticWindow
    }
}